File: lib/parse_html_aam_results.rb
# frozen_string_literal: true

require 'net/http'
require 'zlib'
require 'stringio'
require 'yajl'
require 'pry-byebug'

class ParseHtmlAamResults
  # @param run is a WPT test run for a specific browser
  # @param labelled_tests is a Set of filtered WPT tests
  def self.local_parse(run, labelled_tests)
    uri = URI(run['raw_results_url'])
    # TODO: refactor to actually implement json streaming when moved to Node
    json_io = Net::HTTP.get(uri)
    filtered_results = {}

    # Init parser
    parser = Yajl::Parser.new(symbolize_keys: true)

    # Populate filtered_results and calculate stats when the parser
    # has finished parsing.
    parser.on_parse_complete = lambda do |json|
      json[:results].each do |test_obj|
        # Exclude tests we don't need
        next unless labelled_tests.include?(test_obj[:test]) &&
          test_obj[:test].include?("html-aam") && !test_obj[:test].include?("tentative")

        test_obj[:subtests]&.map do |s|
          # This is trying to infer the element name from the subtest name but
          # there are inconcistencies, especially where the subtest name uses
          # hypens to delinate specifics. In the case where there are hypens in the
          # name, I split again on the hyphen and get the first word which is usually the
          # element name.
          el_name = s[:name].split.first
          if el_name.start_with?("el-")
            el_name.delete_prefix!("el-")
          end
          el_name = el_name.split("-").first

          if filtered_results.has_key?(el_name)
            filtered_results[el_name] << {
              name: s[:name],
              status: s[:status],
              parent_test: test_obj[:test],
              browser: run['browser_name'],
              version: run['browser_version'],
            }
          else
            filtered_results.merge!( {
              el_name => [
                {
                  name: s[:name],
                  status: s[:status],
                  parent_test: test_obj[:test],
                  browser: run['browser_name'],
                  version: run['browser_version'],
                }
              ],
            })
          end
        end

      end

      # Calculate stats
      filtered_results.each do |element_name, subtests|
        passes = subtests.count { |s| s[:status] == "PASS" }
        fails = subtests.count { |s| s[:status] == "FAIL" }


        filtered_results[element_name] << {
          stats: {
            passes: passes,
            fails: fails,
            total_subtests: subtests.count
          }}
      end
    end

    # Parse json; without this the callback above won't run!
    parser.parse(json_io)

    {
      browser: run['browser_name'],
      version: run['browser_version'],
      results: filtered_results
    }
  end
end

File: lib/parseHtmlAamResults.js
'use strict';

const https = require('https');
const { chain } = require('stream-chain');
const { parser } = require('stream-json');
const { pick } = require('stream-json/filters/Pick');
const { streamArray } = require('stream-json/streamers/StreamArray');

/**
 * Downloads a WPT raw results JSON and restructures it, using HTTP and JSON
 * streaming so the full payload is never held in memory at once.
 *
 * @param {Object} run - A WPT run object from the /api/runs endpoint.
 *   Expected keys: raw_results_url, browser_name, browser_version
 * @param {Set<string>} labelledTests - Set of test paths that carry the
 *   target label (e.g. 'a11y'), from the /api/search endpoint.
 * @returns {Promise<{browser: string, version: string, results: Object}>}
 */
function parseHtmlAamResults(run, labelledTests) {
  return new Promise((resolve, reject) => {
    const filteredResults = {};
    const url = new URL(run.raw_results_url);

    https.get(url, (response) => {
      if (response.statusCode !== 200) {
        response.resume(); // drain the socket so it can be reused
        reject(new Error(`HTTP ${response.statusCode} fetching ${run.raw_results_url}`));
        return;
      }

      // chain() pipes each stage into the next and propagates errors.
      //
      // response          – Node.js Readable: raw bytes arrive here as the
      //                     download progresses; nothing is buffered up front.
      // parser()          – stream-json tokeniser: converts the byte stream
      //                     into a stream of JSON tokens.
      // pick(...)         – discards every token that isn't inside the
      //                     top-level "results" key, so we never pay the cost
      //                     of materialising any other part of the document.
      // streamArray()     – assembles each element of the results array into
      //                     a plain JS value and emits { key, value } objects.
      const pipeline = chain([
        response,
        parser(),
        pick({ filter: 'results' }),
        streamArray(),
      ]);

      pipeline.on('data', ({ value: testObj }) => {
        // Apply the same three-part filter as the Ruby implementation:
        //   1. must be in the set of accessibility-labelled tests
        //   2. must live under html-aam
        //   3. must not be a tentative test
        if (
          !labelledTests.has(testObj.test) ||
          !testObj.test.includes('html-aam') ||
          testObj.test.includes('tentative')
        ) {
          return;
        }

        for (const subtest of (testObj.subtests ?? [])) {
          // Infer element name from the subtest name.
          // The subtest name begins with the element name (space-delimited),
          // with an optional "el-" prefix and optional hyphen-separated
          // qualifiers — mirrors the heuristic in parse_html_aam_results.rb.
          let elName = subtest.name.split(' ')[0];
          if (elName.startsWith('el-')) {
            elName = elName.slice(3);
          }
          elName = elName.split('-')[0];

          const subtestData = {
            name: subtest.name,
            status: subtest.status,
            parent_test: testObj.test,
            browser: run.browser_name,
            version: run.browser_version,
          };

          if (filteredResults[elName]) {
            filteredResults[elName].push(subtestData);
          } else {
            filteredResults[elName] = [subtestData];
          }
        }
      });

      pipeline.on('end', () => {
        // Append a stats summary to each element's subtest list.
        for (const subtests of Object.values(filteredResults)) {
          const passes = subtests.filter(s => s.status === 'PASS').length;
          const fails = subtests.filter(s => s.status === 'FAIL').length;
          subtests.push({
            stats: {
              passes,
              fails,
              total_subtests: subtests.length,
            },
          });
        }

        resolve({
          browser: run.browser_name,
          version: run.browser_version,
          results: filteredResults,
        });
      });

      pipeline.on('error', reject);
    }).on('error', reject);
  });
}

module.exports = { parseHtmlAamResults };
